assignment8/sql_intro.py

```diff
@@ -0,0 +1,467 @@
+# Amir Daniali
+# Code the Dream Python 101
+# Week 8
+
+import pathlib
+import sqlite3
+import pandas as pd
+
+
+# Task 3: I wrote classes instead of functions to make the program cleaner
+# and follow OOP encapsulation using singleton/DAO design pattern.
+class ConnectionManager:
+    _instance = None
+
+    def __new__(cls, db_path):
+        """creates a singleton object, if it is not created,
+        else returns the previous singleton object"""
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            cls._instance._init_connection(db_path)
+        return cls._instance
+
+    def _init_connection(self, db_path):
+        self.connection = sqlite3.connect(db_path, check_same_thread=False)
+        self.connection.row_factory = sqlite3.Row
+        self.connection.execute("PRAGMA foreign_keys = 1")
```

Comment:
Nice way to go -- it is easy to forget using PRAGMA when defining foreign keys with sqlite.